compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,263 @@
+};
+
+class DistributeElementwise
+    : public OpTraitRewritePattern<OpTrait::Elementwise> {
+public:
+  DistributeElementwise(MLIRContext *context, VectorLayoutAnalysis &analysis,
+                        VectorLayoutOptions &options,
+                        PatternBenefit benefit = 1)
+      : OpTraitRewritePattern<OpTrait::Elementwise>(context, benefit),
+        analysis(analysis), options(options) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    if (!OpTrait::hasElementwiseMappableTraits(op))
+      return failure();
+
+    // Get the distributed operands.
+    SmallVector<Value> operands;
+    for (Value operand : op->getOperands()) {
+      if (auto vectorOperand = dyn_cast<VectorValue>(operand)) {
+        operand = getDistributed(rewriter, vectorOperand, options);
+      }
+      operands.push_back(operand);
+    }
+
+    // Get the new distributed vector types for the operation.
+    SmallVector<Type> resultTypes;
+    for (Value result : op->getResults()) {
+      Type resultType = result.getType();
+
+      // Distribute vector result types.
+      if (auto vectorResult = dyn_cast<VectorValue>(result)) {
+        resultType = VectorType::get(options.getDistributedShape(vectorResult),
+                                     vectorResult.getType().getElementType());
+      }
+      resultTypes.push_back(resultType);
+    }
+
+    // Replace the original op with the distributed op.
+    Operation *distributedOp =
+        rewriter.create(op->getLoc(), op->getName().getIdentifier(), operands,
+                        resultTypes, op->getAttrs());
+    replaceOpWithDistributedValues(rewriter, op, options,
+                                   distributedOp->getResults());
+    return success();
+  }
+
+private:
+  VectorLayoutAnalysis &analysis;
+  VectorLayoutOptions &options;
+};
+
+/// A rewriter for the pattern rewriting driver.
+class VectorDistributionRewriter : public PatternRewriter,
+                                   RewriterBase::Listener {
+public:
+  VectorDistributionRewriter(MLIRContext *ctx, DenseSet<Operation *> &erasedOps)
+      : PatternRewriter(context, this), erasedOps(erasedOps) {}
+
+  /// We need to keep track of erased operations so that we never try to
+  /// distribute them again.
+  void notifyOperationRemoved(Operation *op) override { erasedOps.insert(op); }
+
+private:
+  // Reference to operations to be erased and the worklist of the driver.
+  DenseSet<Operation *> &erasedOps;
+};
+
+static bool canDistribute(Operation *op, VectorLayoutAnalysis &analysis) {
+  auto values = llvm::to_vector_of<Value>(op->getOperands());
+  llvm::append_range(values, op->getResults());
+
+  // Check if all operands and results of this operation have a layout.
+  return llvm::all_of(values, [&](Value value) -> bool {
+    if (auto vectorValue = dyn_cast<VectorValue>(value)) {
+      return analysis.getLayout<Attribute>(vectorValue) != nullptr;
```

Comment:
the llvm coding style discourages this AFAIR: 
```suggestion
      return analysis.getLayout<Attribute>(vectorValue);
```

Reply:
I did an explicit cast instead (bool)analysis.getLayout... . Not sure if that's preferred more.